qualtran/_infra/bloq.py

```diff
@@ -249,16 +251,73 @@ def add_my_tensors(
-        if you know symbolic expressions for the counts, or if you need to "generalize"
-        the subbloqs by overwriting bloq attributes that do not affect its cost with generic
-        sympy symbols (perhaps with the aid of the provided `SympySymbolAllocator`).
+        This method must return a set of `(bloq, n)` tuples where `bloq` is called `n` times in
+        the decomposition. This method defines one level of the call graph, specifically the
+        edges from this bloq to its immediate children. To get the full graph,
+        call `Bloq.call_graph()`.
+
+        By default, this method will use `self.decompose_bloq()` to count the bloqs called
+        in the decomposition. By overriding this method, you can provide explicit call counts.
+        This is appropriate if: 1) you can't or won't provide a complete decomposition, 2) you
+        know symbolic expressions for the counts, or 3) you need to "generalize" the subbloqs
+        by overwriting bloq attributes that do not affect its cost with generic sympy symbols using
+        the provided `SympySymbolAllocator`.
+        """
+        return self.decompose_bloq().build_call_graph(ssa)
+
+    def call_graph(
+        self,
+        generalizer: Callable[['Bloq'], Optional['Bloq']] = None,
+        keep: Optional[Sequence['Bloq']] = None,
+        max_depth: Optional[int] = None,
+    ) -> Tuple['nx.DiGraph', Dict['Bloq', Union[int, 'sympy.Expr']]]:
+        """Get the bloq call graph and call totals.
+
+        The call graph has edges from a parent bloq to each of the bloqs that it calls in
+        its decomposition. The number of times it is called is stored as an edge attribute.
+        To specify the bloq call counts for a specific node, override `Bloq.build_call_graph()`.
+
+        Args:
+            generalizer: If provided, run this function on each (sub)bloq to replace attributes
+                that do not affect resource estimates with generic sympy symbols. If the function
+                returns `None`, the bloq is omitted from the counts graph.
+            keep: If this function evaluates to True for the current bloq, keep the bloq as a leaf
+                node in the call graph instead of recursing into it.
+            max_depth: If provided, build a call graph with at most this many layers.
+
+        Returns:
+            g: A directed graph where nodes are (generalized) bloqs and edge attribute 'n' reports
+                the number of times successor bloq is called via its predecessor.
+            sigma: Call totals for "leaf" bloqs. We keep a bloq as a leaf in the call graph
+                according to `keep` and `max_depth` (if provided) or if a bloq cannot be
+                decomposed.
+        """
+        from qualtran.resource_counting.bloq_counts import get_bloq_call_graph
+
+        return get_bloq_call_graph(self, generalizer=generalizer, keep=keep, max_depth=max_depth)
+
+    def bloq_counts(
```

Comment:
Also, I see that most of the usages are either in tests or notebooks (or `t_complexity` method, which looks redundant in the first place), so that's further evidence that we can get rid of the method without compromising code readability.

Reply:
you'd need `call_graph(max_depth=1)[1]`

yes, usages are primarily in "user code". This method is to prevent users from calling `build_call_graph` directly. There is a fundamental difference between finding the immediate children bloqs vs the recursive call graph